controllers/pvc_test.go

```diff
@@ -0,0 +1,240 @@
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/cybozu-go/moco/pkg/constants"
+	"github.com/cybozu-go/moco/pkg/metrics"
+	"github.com/google/go-cmp/cmp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	corev1ac "k8s.io/client-go/applyconfigurations/core/v1"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	mocov1beta2 "github.com/cybozu-go/moco/api/v1beta2"
+)
+
+func TestReconcilePVC(t *testing.T) {
+	tests := []struct {
+		name        string
+		cluster     *mocov1beta2.MySQLCluster
+		setupClient func(*testing.T) client.Client
+		wantSize    resource.Quantity
+		wantMetrics string
+	}{
+		{
+			name:    "resize succeeded",
+			cluster: newMySQLClusterWithVolumeSize(resource.MustParse("2Gi")),
+			setupClient: func(t *testing.T) client.Client {
+				cluster := newMySQLClusterWithVolumeSize(resource.MustParse("2Gi"))
+				sts := newStatefulSetWithVolumeSize(resource.MustParse("1Gi"))
+				return setupMockClient(t, cluster, sts)
+			},
+			wantSize: resource.MustParse("2Gi"),
+			wantMetrics: `# HELP moco_cluster_volume_resized_total The number of successful volume resizes
+# TYPE moco_cluster_volume_resized_total counter
+moco_cluster_volume_resized_total{name="mysql-cluster",namespace="default"} 1
+`,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			registry := prometheus.NewRegistry()
+			r := &MySQLClusterReconciler{Client: tt.setupClient(t)}
+
+			metrics.Register(registry)
+
+			err := r.reconcilePVC(ctx, ctrl.Request{NamespacedName: types.NamespacedName{
+				Namespace: tt.cluster.Namespace,
+				Name:      tt.cluster.Name,
+			}}, tt.cluster)
+			if err != nil {
+				t.Fatalf("reconcilePVC() error = %v", err)
+			}
+
+			var pvc corev1.PersistentVolumeClaim
+			if err := r.Get(ctx, types.NamespacedName{Name: "mysql-data-moco-mysql-cluster-0", Namespace: tt.cluster.Namespace}, &pvc); err != nil {
+				t.Fatalf("failed to get PVC: %v", err)
+			}
+			if !pvc.Spec.Resources.Requests.Storage().Equal(tt.wantSize) {
+				t.Errorf("unexpected PVC size: got: %s, want: %s", pvc.Spec.Resources.Requests.Storage().String(), tt.wantSize.String())
+			}
+
+			if err := testutil.GatherAndCompare(registry, strings.NewReader(tt.wantMetrics), "moco_cluster_volume_resized_total"); err != nil {
+				t.Errorf("metrics comparison failed: %v", err)
+			}
+		})
+	}
+}
+
+func setupMockClient(t *testing.T, cluster *mocov1beta2.MySQLCluster, sts *appsv1.StatefulSet) client.Client {
```

Comment:
In the env test, controller-manager is not running, so testing with child resources is not possible (e.g. StatefulSet's child PVCs).
Therefore, the mock client is used.

I will add the E2E test later.